Docs and example programs for the GPU execution feature

ahayashi opened the pull request:
Provide build instructions and example Java programs for the GPU code generation and execution feature in OpenJ9.

keithc-ca commented:
@ahayashi Please see [1] for information about signing the ECA.

[1] https://github.com/eclipse/openj9/blob/master/CONTRIBUTING.md

keithc-ca commented:
Other than the .png file, all will need a copyright notice added.

ahayashi commented on doc/compiler/GPU/README.md:
````diff
@@ -0,0 +1,228 @@
+#### Edit the Dockerfile
+Once you have confirmed that your GPU is recognized and the CUDA compiler ```nvcc``` is available within a docker container, pick one of the JDK Dockerfiles in "Prepare your system" in [the original OpenJ9 build instruction](https://www.eclipse.org/openj9/oj9_build.html). Then, edit the picked Dockerfile to change the base image to GPU-aware one.
+
+For example, if you choose the Dockerfile for [Linux 64-bit (x86_64)](https://github.com/eclipse/openj9/blob/master/buildenv/docker/jdk8/x86_64/ubuntu16/Dockerfile), look for ```FROM``` and update the line like this:
````
Sure, I will create GPU-aware Docker files. Do you expect that those Dockerfiles are committed here (```doc/compiler/GPU```) or more appropriate directories such as ```/buildenv/docker/jdk...```?

gita-omr replied:
@keithc-ca thanks for the comments. Just curious, which Docker file do you use to build VM-based GPU support? (eg. for java.util.Arrays.sort())

keithc-ca replied:
I don't normally use docker to do that.

FYI, CUDA-enabled builds should soon be available at https://adoptopenjdk.net/nightly.html?jvmVariant=openj9 now that AdoptOpenJDK/openjdk-build#1187 is merged.

ahayashi commented on doc/compiler/GPU/README.md:
```diff
@@ -0,0 +1,228 @@
+RUN ln -sf g++ /usr/bin/c++ \
+  && ln -sf g++-4.8 /usr/bin/g++ \
+  && ln -sf gcc /usr/bin/cc \
+  && ln -sf gcc-4.8 /usr/bin/gcc
```
I'll incorporate this into new Dockerfiles.

ahayashi commented on doc/compiler/GPU/README.md:
````diff
@@ -0,0 +1,228 @@
+Now you're ready to build OpenJDK with OpenJ9 with the GPU code generation capability enabled. When doing ```make```, you essentially need to give ```ENABLE_GPU=1``` and the path to ```CUDA_HOME``` and ```GDK_HOME```:
+
+```
+make all ENABLE_GPU=1 CUDA_HOME=/usr/local/cuda GDK_HOME=/usr/include/nvidia/gdk
````
Sure, I updated the text, thanks.

gita-omr replied:
@keithc-ca does --enable-cuda defines -DENABLE_GPU for JIT?

keithc-ca replied:
> @keithc-ca does --enable-cuda define -DENABLE_GPU for JIT?

Yes. Here's an excerpt from the log of a build I did this morning:
```
make[5]: Entering directory '/home/keithc/space/jdk11/build/linux-x86_64-normal-server-release/vm/compiler'
J9_VERSION:           29
TR_BUILD_NAME:        7f06832
J9SRC:                /home/keithc/space/jdk11/build/linux-x86_64-normal-server-release/vm/compiler/..
JIT_SRCBASE:          /home/keithc/space/jdk11/build/linux-x86_64-normal-server-release/vm/compiler/..
JIT_OBJBASE:          /home/keithc/space/jdk11/build/linux-x86_64-normal-server-release/vm/compiler/../objs
JIT_DLL_DIR:          /home/keithc/space/jdk11/build/linux-x86_64-normal-server-release/vm/compiler/..
OMR_DIR:              /home/keithc/space/jdk11/build/linux-x86_64-normal-server-release/vm/compiler/../omr
BUILD_CONFIG:         prod
NUMBER_OF_PROCESSORS: 
VERSION_MAJOR:        11
ENABLE_GPU:           1
CUDA_HOME:            /home/keithc/opt/cuda-9.2
GDK_HOME:             /usr/include/nvidia/gdk
```

ahayashi commented:
Thanks, @keithc-ca! I'll incorporate all the feedback. One question, re. copyright notice, should I also add it to .md files?

keithc-ca commented on doc/compiler/GPU/README.md:
````diff
@@ -0,0 +1,228 @@
+
+Before proceeding to the next step, it is a good idea to make sure the installations are successfull. Let's make sure that your GPUs are recognized and the CUDA compiler ```nvcc``` is available within a docker container. In the following example, we create a docker container based on the ```nvidia/cuda:9.2-devel-ubuntu16.04``` image, and perform two commands: 1) the ```nvidia-smi``` command, whic ...
+
+**Note:** you can pick any CUDA version that is supported in your GPU, but always make sure you pick a ```devel``` image because the GPU code generation feature requires the CUDA compiler toolchain. A full list of the available CUDA docker images can be found at [the Docker Hub](https://hub.docker.com/r/nvidia/cuda/). For ```ppc64le``` users, see [here](https://hub.docker.com/r/nvidia/cuda-ppc64l ...
````
For the scenarios discussed here, I don't think you will actually be using `nvcc`. Only the CUDA toolkit header files are required for building the JVM. At runtime, only the toolkit and driver shared libraries are needed; JIT generates IR code for `nvvm` which transforms it into GPU code. The code underlying the acceleration of `Arrays.sort()` is generated as PTX code which can be loaded directly by the driver.

ahayashi replied:
Right, here I used ```nvcc``` only for checking purposes. I'll update the text just to check required things such as ```nvvm.h``` and so on.

ahayashi commented:
>Yes, see [1] for an example.
>[1] https://raw.githubusercontent.com/eclipse/openj9/master/doc/README.md
@keithc-ca, great, thanks!

fjeremic commented on doc/compiler/GPU/README.md:
````diff
@@ -0,0 +1,228 @@
+
+The ```heavy()``` method contains Java's Parallel Stream API and is at first interpreted on the JVM. The JVM periodically monitors the *hotness* of the method, and, when needed, generates an optimized native CPU program. The JIT compiler at first performs a limited set of compiler optimizations over intermediate representation (IR) so as not to hinder the main program execution. However, dependin ...
+
+![An overview of the GPU code generation.](./figs/gpucodegen.png "An overview of the GPU code generation." )
````
We prefer not to check in binary files into the repository, otherwise things start to bloat very quick for everyone cloning the repository. For images such as this one we can use ASCII art however. I've taken your image and created an ASCII version which we can use instead of checking in the binary PNG:

```
   +----------------+
   | Java Bytecode  |
   +----------------+
           |
+----------|---------------------------------------------------------------------------------------+
|          v                                                                                       |
|  +----------------+       +----------------+       +----------------+       +----------------+   |   +-----------+
|  |  Translation   | +---> |    Existing    | +---> |Parallel Streams| +---> |Target Machine  | +---> |    CPU    |
|  |   to our IR    |       | Optimizations  |       | Identification |   +-> |Code Generation |   |   |   Binary  |
|  +----------------+       +----------------+       +----------------+   |   +----------------+   |   +-----------+
|                                                            |            |                        |   |GPU Runtime|
|                                                            v            |                        |   +-----------+
|                                                    +----------------+   |                        |
|                                                    |  Optimization  +---+                        |
|                                                    |    for GPUs    |                            |
| Our JIT Compiler                                   +----------------+                            |
|                                                            |                                     |
+--------------------------------------------------------------------------------------------------+
                                                             |
+------------------------------------------------------------v-------------------------------------+
|                                                    +----------------+       +----------------+   |   +-----------+
|                                                    |    NVVM IR     | +---> |    libNVVM     | +---> |    GPU    |
| NVIDIA's Tool Chain                                |                |       |                |   |   |   Binary  |
|                                                    +----------------+       +----------------+   |   +-----------+
+--------------------------------------------------------------------------------------------------+

```

ahayashi replied:
@fjeremic thank you very much! I'll definitely use the ASCII version.

fjeremic replied:
For reference I've created the above using ASCIIFlow [1] which is a very handy tool. Thanks for your contribution!

[1] http://asciiflow.com/

ahayashi commented:
Hi all,

I've incorporated most of the comments and updated the files. Also, it's worth noting that I've added the copyright notice to all the files.

@fjeremic I've removed the png file and put the ASCII version in the README file, thanks again!

@keithc-ca @gita-omr one thing we need to discuss is how we prepare GPU-aware Dockefiles. In short, because the difference between the original dockerfiles and ours are very small, I'm not sure if it's worth preparing for a dedicated Dockerfile for each platform, jdk version, OS, CUDA version. 

Essentially the changes are like this:
1) ```FROM X:Y``` to ```FROM nvidia/cuda:Z-devel-XY```, where X is the name of OS (e.g., ubuntu), Y is the OS version (e.g., 16.04), Z is the CUDA version (e.g., 9.2)
2) adding the ```-f``` option for the ```ln``` commands (Because the CUDA images sometimes already have symbolic links to ```gcc```, ```g++``` in ```/usr/bin```).

For more details, please see below:

__Details of the Dockerfile thing:__ The thing is that we basically use Nvidia's docker hub to get a base image, in which CUDA versions are hardcoded like this: nvidia/cuda:9.2. And the hardcoded docker image name should be put at the beginning of Dockerfile (like FROM nvidia/cuda:9.2-devel-ubuntu16.04). I don't think we want to prepare a dedicated Dockerfile for each platform, jdk version, OS, CUDA version like this:
```
openj9/buildenv/docker/jdk8/x86_64/ubuntu16/Dockerfile // original, FROM ubuntu:16.04 
openj9/buildenv/docker/jdk8/x86_64/ubuntu16/Dockerfile.cuda9 // CUDA 9, FROM nvidia/cuda:9.2-devel-ubuntu16.04
openj9/buildenv/docker/jdk8/x86_64/ubuntu16/Dockerfile.cuda10  // CUDA10, FROM nvidia/cuda:10.1-devel-ubuntu16.04
...
openj9/buildenv/docker/jdk8/x86_64/ubuntu18/Dockerfile // original, FROM ubuntu:18.04
openj9/buildenv/docker/jdk8/x86_64/ubuntu18/Dockerfile.cuda9 // CUDA 9, FROM nvidia/cuda:9.2-devel-ubuntu18.04
openj9/buildenv/docker/jdk8/x86_64/ubuntu18/Dockerfile.cuda10  // CUDA10, FROM nvidia/cuda:10.1-devel-ubuntu18.04
openj9/buildenv/docker/jdk8/ppc64/...

openj9/buildenv/docker/jdk9/...
openj9/buildenv/docker/jdk10/...
openj9/buildenv/docker/jdk11/...
```

At least, I'd like to fix a CUDA version and the updated README.md assumes that users modify the CUDA version in Dockerfile if they want to do so.  However, any suggestions are welcome.

gita-omr commented:
Reviewers, please take a second look and close your reviews.

ahayashi commented:
Let me give you some options for the Dockerfile thing:

Option 1: Provide dedicated Dockerfiles for each OS, OS-version, CUDA-version, platform, ... -> a lot of Dockerfiles
Option 2: Same as option 1, but we fix a CUDA version (say 9.2), to reduce the number of Dockerfiles and let the user update the CUDA version manually.
Option 3: Let the user do the changes (change 1 and 2 in my previous comment) -> we don't need to create dedicated Dockerfiles for GPU containers
Option 4: For the ```FROM``` field add a command-line argument so the user can give their configuration when building the Dockerfile to reduce the number of Dockerfiles  -> we don't need to create dedicated Dockerfiles for GPU containers, but I'm not sure if this is good practice. 

```
// Dockerfile (Option 4)
ARG baseimage=ubuntu:18.04
FROM ${baseimage}

// command line (Option 4)
$ docker build --build-arg baseimage=nvidia/cuda:9.2-devel-ubuntu18.04
```

Please let me know if you have any suggestions. 

Thanks,

Akihiro

budgidiere commented:
Any word on this?

keithc-ca commented:
@ahayashi Do you have plans to address the feedback?

keithc-ca commented:
No response from author: closing.
